File: src/examples/packet_in/packet_in.c
/*
 * Dumps packet-in message.
 *
 * Author: Yasuhito Takamiya <yasuhito@gmail.com>
 *
 * Copyright (C) 2008-2013 NEC Corporation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */


#include <inttypes.h>
#include "trema.h"


static void
handle_packet_in( uint64_t datapath_id, packet_in message ) {
  info( "received a packet_in" );
  info( "datapath_id: %#" PRIx64, datapath_id );
  info( "transaction_id: %#x", message.transaction_id );
  info( "buffer_id: %#x", message.buffer_id );
  info( "total_len: %u", message.total_len );
  info( "reason: %#x", message.reason );
  info( "table_id: %#x", message.table_id );
  info( "cookie: %#" PRIx64, message.cookie );
  char match_string[ MATCH_STRING_LENGTH ];
  memset( match_string, '\0', MATCH_STRING_LENGTH );
  match_to_string( message.match, match_string, sizeof( match_string ) );
  info( "match: %s", match_string );
  info( "data:" );
  dump_buffer( message.data, info );
}


static void
handle_switch_ready( uint64_t datapath_id, void *user_data ) {
  UNUSED( user_data );

  openflow_actions *actions = create_actions();
  append_action_output( actions, OFPP_CONTROLLER, OFPCML_NO_BUFFER );
  openflow_instructions *insts = create_instructions();
  append_instructions_apply_actions( insts, actions );

  buffer *flow_mod = create_flow_mod(
    get_transaction_id(),
    get_cookie(),
    0,
    0,
    OFPFC_ADD,
    0,
    0,
    OFP_LOW_PRIORITY,
    OFP_NO_BUFFER,
    0,
    0,
    OFPFF_SEND_FLOW_REM,
    NULL,
    insts
  );
  send_openflow_message( datapath_id, flow_mod );
  free_buffer( flow_mod );

  delete_instructions( insts );
  delete_actions( actions );
}


int
main( int argc, char *argv[] ) {
  init_trema( &argc, &argv );
  set_packet_in_handler( handle_packet_in, NULL );
  set_switch_ready_handler( handle_switch_ready, NULL );
  start_trema();
  return 0;
}


/*
 * Local variables:
 * c-basic-offset: 2
 * indent-tabs-mode: nil
 * End:
 */

File: src/examples/packet_in/packet-in.rb
#
# Dumps packet-in message.
#
# Author: Yasuhito Takamiya <yasuhito@gmail.com>
#
# Copyright (C) 2008-2013 NEC Corporation
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License, version 2, as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#

class PacketinDumper < Controller
  def switch_ready(datapath_id)
    action = SendOutPort.new(port_number: OFPP_CONTROLLER, max_len: OFPCML_NO_BUFFER)
    apply_ins = ApplyAction.new(actions: [action])
    send_flow_mod_add(datapath_id,
                      priority: OFP_LOW_PRIORITY,
                      buffer_id: OFP_NO_BUFFER,
                      instructions: [apply_ins])
  end

  def packet_in(datapath_id, event)
    puts 'received a packet_in'
    info "datapath_id: #{ datapath_id.to_hex }"
    info "transaction_id: #{ event.transaction_id.to_hex }"
    info "buffer_id: #{ event.buffer_id.to_hex }"
    info "total_len: #{ event.total_len }"
    info "reason: #{ event.reason.to_hex }"
    info "table_id: #{ event.table_id }"
    info "cookie: #{ event.cookie.to_hex }"
    info "in_port: #{ event.match.in_port }"
    info "data: #{ event.data.map! { | byte | '0x%02x' % byte } }"
    info 'packet_info:'
    info "  eth_src: #{ event.eth_src }"
    info "  eth_dst: #{ event.eth_src }"
    info "  eth_type: #{ event.eth_type.to_hex }"

    if event.eth_type == 0x800 || event.eth_type == 0x86dd
      info "  ip_dscp: #{ event.ip_dscp }"
      info "  ip_ecn: #{ event.ip_ecn }"
      info "  ip_proto: #{ event.ip_proto }"
    end

    if event.vtag?
      info "  vlan_vid: #{ event.vlan_vid.to_hex }"
      info "  vlan_prio: #{ event.vlan_prio.to_hex }"
      info "  vlan_tpid: #{ event.vlan_tpid.to_hex }"
      info "  vlan_tci: #{ event.vlan_tci.to_hex }"
    end

    if event.ipv4?
      info "  ipv4_src: #{ event.ipv4_src }"
      info "  ipv4_dst: #{ event.ipv4_dst }"
    end

    if event.ipv6?
      info "  ipv6_src: #{ event.ipv6_src }"
      info "  ipv6_dst: #{ event.ipv6_dst }"
      info "  ipv6_flabel: #{ event.ipv6_flabel.to_hex }"
      info "  ipv6_exthdr: #{ event.ipv6_exthdr.to_hex }"
    end

    if event.arp?
      info "  arp_op: #{ event.arp_op }"
      info "  arp_sha: #{ event.arp_sha }"
      info "  arp_spa: #{ event.arp_spa }"
      info "  arp_tpa: #{ event.arp_tpa }"
    end

    if event.icmpv4?
      info "  icmpv4_type: #{ event.icmpv4_type.to_hex }"
      info "  icmpv4_code: #{ event.icmpv4_code.to_hex }"
    end

    if event.icmpv6?
      info "  icmpv6_type: #{ event.icmpv6_type.to_hex }"
      info "  icmpv6_code: #{ event.icmpv6_code.to_hex }"
    end

    if event.udp?
      info "  udp_src: #{ event.udp_src.to_hex }"
      info "  udp dst: #{ event.udp_dst.to_hex }"
    end

    if event.sctp?
      info "  sctp_src: #{ event.sctp_src.to_hex }"
      info "  sctp_dst: #{ event.sctp_dst.to_hex }"
    end

    if event.pbb?
      info "  pbb_isid: #{ event.pbb_isid.to_hex }"
    end

    if event.mpls?
      info "  mpls_label: #{ event.mpls_label.to_hex }"
      info "  mpls_tc: #{ event.mpls_tc.to_hex }"
      info "  mpls_bos: #{ event.mpls_bos.to_hex }"
    end
  end
end

### Local variables:
### mode: Ruby
### coding: utf-8
### indent-tabs-mode: nil
### End:
